v2/apierror/api_error.go

```diff
@@ -0,0 +1,236 @@
+		var f []string
+		var desc []string
+		for _, x := range v {
+			f = append(f, x.GetField())
+			desc = append(desc, x.GetDescription())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = BadRequest field = %s desc = %s\n",
+			strings.Join(f, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.PreconditionFailure != nil {
+		v := a.details.PreconditionFailure.GetViolations()
+		var t []string
+		var s []string
+		var desc []string
+		for _, x := range v {
+			t = append(t, x.GetType())
+			s = append(s, x.GetSubject())
+			desc = append(desc, x.GetDescription())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = PreconditionFailure type = %s subj = %s desc = %s\n", strings.Join(t, " "),
+			strings.Join(s, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.QuotaFailure != nil {
+		v := a.details.QuotaFailure.GetViolations()
+		var s []string
+		var desc []string
+		for _, x := range v {
+			s = append(s, x.GetSubject())
+			desc = append(desc, x.GetDescription())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = QuotaFailure subj = %s desc = %s\n",
+			strings.Join(s, " "), strings.Join(desc, " ")))
+	}
+
+	if a.details.RequestInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = RequestInfo id = %s data = %s\n",
+			a.details.RequestInfo.GetRequestId(), a.details.RequestInfo.GetServingData()))
+	}
+
+	if a.details.ResourceInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = ResourceInfo type = %s resourcename = %s owner = %s desc = %s\n",
+			a.details.ResourceInfo.GetResourceType(), a.details.ResourceInfo.GetResourceName(),
+			a.details.ResourceInfo.GetOwner(), a.details.ResourceInfo.GetDescription()))
+
+	}
+	if a.details.RetryInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: retry in %s\n", a.details.RetryInfo.GetRetryDelay().AsDuration()))
+
+	}
+	if a.details.Unknown != nil {
+		var s []string
+		for _, x := range a.details.Unknown {
+			s = append(s, fmt.Sprintf("%v", x))
+		}
+		d.WriteString(fmt.Sprintf("error details: name = Unknown  desc = %s\n", strings.Join(s, " ")))
+	}
+
+	if a.details.DebugInfo != nil {
+		d.WriteString(fmt.Sprintf("error details: name = DebugInfo detail = %s stack = %s\n", a.details.DebugInfo.GetDetail(),
+			strings.Join(a.details.DebugInfo.GetStackEntries(), " ")))
+	}
+	if a.details.Help != nil {
+		var desc []string
+		var url []string
+		for _, x := range a.details.Help.Links {
+			desc = append(desc, x.GetDescription())
+			url = append(url, x.GetUrl())
+		}
+		d.WriteString(fmt.Sprintf("error details: name = Help desc = %s url = %s\n",
+			strings.Join(desc, " "), strings.Join(url, " ")))
+	}
+	if a.details.LocalizedMessage != nil {
+		d.WriteString(fmt.Sprintf("error details: name = LocalizedMessge locale = %s msg = %s\n",
+			a.details.LocalizedMessage.GetLocale(), a.details.LocalizedMessage.GetMessage()))
+	}
+	return strings.TrimSpace(d.String())
+}
+
+// GRPCStatus extracts the underlying gRPC Status error.
```

Comment:
I think code mentioned this before, but could you put a comment here that indicates `This method is necessary to fulfill the interface described in https://pkg.go.dev/google.golang.org/grpc/status#FromError.`

Reply:
done!